caching-attribute-service-client/src/main/java/org/hypertrace/core/attribute/service/cachingclient/DefaultCachingAttributeClient.java

```diff
@@ -0,0 +1,87 @@
+package org.hypertrace.core.attribute.service.cachingclient;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableTable;
+import com.google.common.collect.Table;
+import io.grpc.CallCredentials;
+import io.grpc.Channel;
+import io.reactivex.rxjava3.core.Single;
+import java.time.Duration;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadata;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadataFilter;
+import org.hypertrace.core.attribute.service.v1.AttributeServiceGrpc;
+import org.hypertrace.core.attribute.service.v1.AttributeServiceGrpc.AttributeServiceStub;
+
+class DefaultCachingAttributeClient implements CachingAttributeClient {
+
+  private final AttributeServiceStub attributeServiceClient;
+  private final LoadingCache<
+          AttributeCacheContextKey, Single<Table<String, String, AttributeMetadata>>>
+      cache;
+  private final AttributeMetadataFilter attributeFilter;
+
+  DefaultCachingAttributeClient(
+      @Nonnull Channel channel,
+      @Nonnull CallCredentials credentials,
+      int maxCacheContexts,
+      @Nonnull Duration cacheExpiration,
+      @Nonnull AttributeMetadataFilter attributeFilter) {
+
+    this.attributeFilter = attributeFilter;
+    this.attributeServiceClient =
+        AttributeServiceGrpc.newStub(channel).withCallCredentials(credentials);
+    this.cache =
+        CacheBuilder.newBuilder()
+            .maximumSize(maxCacheContexts)
+            .expireAfterWrite(cacheExpiration)
+            .build(CacheLoader.from(this::loadTable));
+  }
+
+  @Override
+  public Single<AttributeMetadata> get(String scope, String key) {
+    return this.getOrInvalidate(AttributeCacheContextKey.forCurrentContext())
```

Comment:
Curious, why are we invalidating the cache when the entry is missing? Let's say you're looking for a non-existing attribute, would every following call go over the network?

Reply:
Yeah good discussion point. My thinking here was that a get call _should_ always work. If we are using a known key and scope and it is missing, is it because attributes have been recently updated and the cache is out of sync? Should we give an error without verifying that our state is correct? If it were invalid, that call would go over the network. Subsequent calls wouldn't unless they were for that invalid key. In other words, this behavior is basically saying only cache successes, make a remote call before marking failure so we know all failures are valid.

I could definitely be convinced to change this, I'm actually leaning in the other direction now as a non existent attribute might be requested many times in sequence.

Reply:
I think the case where we should definitely avoid querying over and over is, an invalid attribute is looked up and the call to server has succeeded but didn't return any result. That's a clear case that Cache can remember and return empty response next time when same invalid key is looked up (until cache expiry).

Reply:
I'm so incredibly dumb. I removed this and spent like an hour trying to fix the test only to finally realize my test was right all along (saved the day!) and I had apparently forgotten why I originally wrote this and explained it completely wrong. So, ignore what I said above. What this is _actually_ doing is not invalidating when an entry is missing (that error is generated on line 51 and won't invalidate the cache), but retrieving the async table from the cache. If that async call had ended up in an error (for example, the downstream was unavailable) we don't want to cache that error state but retry. Once a call succeeds, even if it returned an empty table, we won't manually invalidate it.